protobuf/grpc_service.proto

```diff
@@ -1477,3 +1486,74 @@ message CudaSharedMemoryUnregisterRequest
 //@@   Response message for CudaSharedMemoryUnregister.
 //@@
 message CudaSharedMemoryUnregisterResponse {}
+
+//@@
+//@@.. cpp:var:: message TraceRequest
+//@@
+//@@   Request message for Trace.
+//@@
+message TraceRequest
+{
+  //@@
+  //@@  .. cpp:var:: message SettingValue
+  //@@
+  //@@     The values to be associated with a trace setting.
+  //@@
+  message SettingValue
+  {
+    //@@
+    //@@    .. cpp:var:: string value (repeated)
+    //@@
+    //@@       The value.
+    //@@
+    repeated string value = 3;
+  }
+
+  //@@  .. cpp:var:: map<string,SettingValue> settings
+  //@@
+  //@@     The new setting values to be updated,
+  //@@     settings that are not specified will remain unchanged.
+  //@@
+  map<string, SettingValue> settings = 1;
+
+  //@@
+  //@@  .. cpp:var:: string name
+  //@@
+  //@@     The name of the model to apply the new trace settings.
+  //@@     If not given, the new settings will be applied globally.
+  //@@
+  string model_name = 2;
+
+  //@@
+  //@@  .. cpp:var:: bool clear_setting
+  //@@
+  //@@     Clear the current model specific setting if specified, and
+  //@@     any new setting values will not be applied.
+  //@@     This option will be ignored if 'model_name' is not given.
+  //@@
+  bool clear_setting = 3;
```

Comment:
Why is this needed. Can we instead just say that if SettingValue:value is empty then the setting is cleared. If user doesn't want to change a value then they can just not provide it in 'settings'.

Reply:
Different from HTTP where we can use GET / POST to separate whether the setting will be changed, in GRPC, an empty `settings` is ambiguous on whether the client want to clear the settings or to only retrieve the current setting.

I need to check on whether map has `has_xxx` API in proto3, if so, we may utilize that to only look up values if `has_settings == false` and a clear will be done if `has_settings == true && settings.size() == 0`

Reply:
I think it is ok for GRPC to act like HTTP in that the response should always return all the global or model settings, no matter what (if anything) is specified in the request. Then anything specified in the request is considered an update to the setting.